src/utils/InteractiveChildProcess.ts

```diff
@@ -0,0 +1,169 @@
+
+export interface IInteractiveChildProcessOptions {
+    command: string;
+    args: string[];
+    outputChannel?: vscode.OutputChannel;
+    workingDirectory?: string;
+    showTimeInOutputChannel?: boolean;
+    outputFilterSearch?: RegExp;
+    outputFilterReplace?: string;
+}
+
+export class InteractiveChildProcess {
+    private _childProc: cp.ChildProcess;
+    private readonly _options: IInteractiveChildProcessOptions;
+    private _startTime: number;
+    private _error: unknown;
+    private _isKilling: boolean;
+
+    private constructor(options: IInteractiveChildProcessOptions) {
+        this._options = options;
+    }
+
+    private readonly _onStdOutEmitter: EventEmitter<string> = new EventEmitter<string>();
+    public readonly onStdOut: Event<string> = this._onStdOutEmitter.event;
+
+    private readonly _onStdErrEmitter: EventEmitter<string> = new EventEmitter<string>();
+    public readonly onStdErr: Event<string> = this._onStdErrEmitter.event;
+
+    private readonly _onErrorEmitter: EventEmitter<unknown> = new EventEmitter<unknown>();
+    public readonly onError: Event<unknown> = this._onErrorEmitter.event;
+
+    public static async create(options: IInteractiveChildProcessOptions): Promise<InteractiveChildProcess> {
+        let child: InteractiveChildProcess = new InteractiveChildProcess(options);
+        await child.startCore();
+        return child;
+    }
+
+    public kill(): void {
+        this._isKilling = true;
+        this._childProc.kill();
+    }
+
+    public writeLine(text: string): void {
+        this.writeLineToOutputChannel(text, stdInPrefix);
+        this._childProc.stdin.write(text + os.EOL);
+    }
+
+    private async startCore(): Promise<void> {
+        this._startTime = Date.now();
+        const formattedArgs: string = this._options.args.join(' ');
+
+        let workingDirectory = this._options.workingDirectory || os.tmpdir();
+        const options: cp.SpawnOptions = {
+            cwd: workingDirectory,
+
+            // Using shell=true would mean that we can pass paths that will be resolved by the shell, but since
+            //   the command is run in the shell, handling errors (such as command not found) would be more indirect,
+            //   coming through STDERR instead of the error event
+            shell: false
+        };
+
+        this.writeLineToOutputChannel(`Starting executable: "${this._options.command}" ${formattedArgs}`);
+        this._childProc = cp.spawn(this._options.command, this._options.args, options);
+
+        this._childProc.stdout.on('data', (data: string | Buffer) => {
+            let text = data.toString();
+            this.writeLineToOutputChannel(text);
+            this._onStdOutEmitter.fire(text);
+        });
+
+        this._childProc.stderr.on('data', (data: string | Buffer) => {
+            let text = data.toString();
+            this._onStdErrEmitter.fire(text);
+            this.writeLineToOutputChannel(text, stdErrPrefix);
```

Comment:
Why is the order of firing & writing to output different between stdout's and stderr's  handling of data?